File: frontend/src/types/plan.ts
/** Plan and Subscription domain types — mirrors backend entities */

export interface Plan {
  id: string
  name: string
  displayName: string
  priceCents: number
  maxActiveAlerts: number | null
  minFrequencyMinutes: number
  priceHistoryDays: number | null
  maxSources: number | null
  hasPushNotifications: boolean
  hasWhatsappNotifications: boolean
  hasApiAccess: boolean
  isActive: boolean
}

export interface Subscription {
  id: string
  userId: number
  planId: number
  planName: string
  status: 'active' | 'canceled' | 'past_due'
  currentPeriodStart?: string | null
  currentPeriodEnd?: string | null
  canceledAt?: string | null
  createdAt?: string
  updatedAt?: string
}

export interface PlanLimits {
  planName: string
  maxActiveAlerts: number | null
  minFrequencyMinutes: number
  priceHistoryDays: number | null
  maxSources: number | null
}

File: backend/src/app/entities/plan.py
from pydantic import BaseModel, ConfigDict


class Plan(BaseModel):
    """Plan domain entity representing a subscription tier."""

    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    name: str  # "free", "pro", "business"
    display_name: str  # "Free", "Pro", "Business"
    price_cents: int = 0  # price in BRL cents (e.g. 2900 = R$29)
    max_active_alerts: int | None = None  # None = unlimited
    min_frequency_minutes: int = 360  # minimum check interval
    price_history_days: int | None = 7  # None = unlimited
    max_sources: int | None = None  # None = all
    has_push_notifications: bool = False
    has_whatsapp_notifications: bool = False
    has_api_access: bool = False
    is_active: bool = True
